include/vara/Feature/ConstraintParser.h

```diff
@@ -0,0 +1,353 @@
+//                               ConstraintToken
+//===----------------------------------------------------------------------===//
+
+class ConstraintToken {
+public:
+  using Precedence = int;
+
+  enum class ConstraintTokenKind {
+    ERROR,
+    END_OF_FILE,
+    WHITESPACE,
+    IDENTIFIER,
+    NUMBER,
+    L_PAR,
+    R_PAR,
+    NOT,
+    OR,
+    AND,
+    EQUAL,
+    NOT_EQUAL,
+    LESS,
+    GREATER,
+    IMPLIES,
+    EQUIVALENT,
+    PLUS,
+    MINUS,
+    STAR,
+    NEG
+  };
+
+  ConstraintToken(ConstraintTokenKind Kind) : Kind(Kind) {}
+  ConstraintToken(ConstraintTokenKind Kind, std::string Value)
+      : Kind(Kind), Value(Value) {}
+  ConstraintToken(const ConstraintToken &) = delete;
+  ConstraintToken &operator=(ConstraintToken &) = delete;
+  ConstraintToken(ConstraintToken &&) = delete;
+  ConstraintToken &operator=(ConstraintToken &&) = delete;
+  virtual ~ConstraintToken() = default;
+
+  [[nodiscard]] ConstraintTokenKind getKind() const { return Kind; };
+
+  [[nodiscard]] std::optional<std::string> getValue() { return Value; }
+
+  [[nodiscard]] Precedence calcPrecedence() const {
+    switch (Kind) {
+    case ConstraintTokenKind::STAR:
+      return 1;
+    case ConstraintTokenKind::PLUS:
+    case ConstraintTokenKind::MINUS:
+      return 2;
+    case ConstraintTokenKind::LESS:
+    case ConstraintTokenKind::GREATER:
+      return 3;
+    case ConstraintTokenKind::EQUAL:
+    case ConstraintTokenKind::NOT_EQUAL:
+      return 4;
+    case ConstraintTokenKind::AND:
+      return 5;
+    case ConstraintTokenKind::OR:
+      return 6;
+    default:
+      return 7;
+    }
+  }
+
+private:
+  const ConstraintTokenKind Kind;
+  const std::optional<std::string> Value{std::nullopt};
+};
+
+//===----------------------------------------------------------------------===//
+//                               ConstraintToken
+//===----------------------------------------------------------------------===//
+
+class ConstraintLexer {
+public:
+  using TokenListTy = std::vector<std::unique_ptr<ConstraintToken>>;
+
+  explicit ConstraintLexer(std::string Cnt) : Cnt(std::move(Cnt)) {}
+
+  TokenListTy buildTokenList() {
+    TokenListTy TokenList;
+    for (int Pos = 0; Pos < Cnt.size();) {
+      auto Result = munch(Pos);
+      TokenList.push_back(std::move(Result.first));
+      Pos += Result.second;
+      if (auto Kind = TokenList.back()->getKind();
+          Kind == ConstraintToken::ConstraintTokenKind::END_OF_FILE ||
+          Kind == ConstraintToken::ConstraintTokenKind::ERROR) {
+        break;
+      }
+    }
+
+    // ensure last token is always either EOF or an error
+    if (auto Kind = TokenList.back()->getKind();
+        Kind != ConstraintToken::ConstraintTokenKind::END_OF_FILE &&
+        Kind != ConstraintToken::ConstraintTokenKind::ERROR) {
+      TokenList.push_back(std::make_unique<ConstraintToken>(
+          ConstraintToken::ConstraintTokenKind::END_OF_FILE));
+    }
+
+    return std::move(TokenList);
```

Comment:
```suggestion
    return TokenList;
```
If all return paths return the same local stack value we don't need to move it, RVO will take care of everything :smile: .